packages/node-core/src/indexer/project.service.spec.ts

```diff
@@ -0,0 +1,167 @@
+      {getDynamicDatasources: jest.fn()} as unknown as DynamicDsService<any>,
+      null as unknown as any,
+      null as unknown as any
+    );
+  });
+
+  it('hasDataSourcesAfterHeight', async () => {
+    service.getDataSources = jest.fn().mockResolvedValue([{endBlock: 100}, {endBlock: 200}, {endBlock: 300}]);
+
+    const result = await service.hasDataSourcesAfterHeight(150);
+    expect(result).toBe(true);
+  });
+
+  it('hasDataSourcesAfterHeight - undefined endBlock', async () => {
+    service.getDataSources = jest.fn().mockResolvedValue([{endBlock: 100}, {endBlock: undefined}]);
+
+    const result = await service.hasDataSourcesAfterHeight(150);
+    expect(result).toBe(true);
+  });
+
+  it('getDataSources', async () => {
+    (service as any).project.dataSources = [
+      {startBlock: 100, endBlock: 200},
+      {startBlock: 1, endBlock: 100},
+    ];
+    (service as any).dynamicDsService.getDynamicDatasources = jest
+      .fn()
+      .mockResolvedValue([{startBlock: 150, endBlock: 250}]);
+
+    const result = await service.getDataSources(175);
+    expect(result).toEqual([
+      {startBlock: 100, endBlock: 200},
+      {startBlock: 150, endBlock: 250},
+    ]);
+  });
+
+  describe('getDatasourceMap', () => {
+    it('should add endBlock heights correctly', () => {
+      (service as any).dynamicDsService.dynamicDatasources = [];
+      (service as any).projectUpgradeService = {
+        projects: [
+          [
+            1,
+            {
+              dataSources: [
+                {startBlock: 1, endBlock: 300},
+                {startBlock: 10, endBlock: 20},
+                {startBlock: 1, endBlock: 100},
+                {startBlock: 50, endBlock: 200},
+                {startBlock: 500},
+              ],
+            },
+          ],
+        ],
+      } as any;
+
+      const result = service.getDataSourcesMap();
+      expect(result.getAll()).toEqual(
+        new Map([
+          [
+            1,
+            [
+              {startBlock: 1, endBlock: 300},
+              {startBlock: 1, endBlock: 100},
+            ],
+          ],
+          [
+            10,
+            [
+              {startBlock: 1, endBlock: 300},
+              {startBlock: 1, endBlock: 100},
+              {startBlock: 10, endBlock: 20},
+            ],
+          ],
+          [
+            21,
+            [
+              {startBlock: 1, endBlock: 300},
+              {startBlock: 1, endBlock: 100},
+            ],
+          ],
+          [
+            50,
+            [
+              {startBlock: 1, endBlock: 300},
+              {startBlock: 1, endBlock: 100},
+              {startBlock: 50, endBlock: 200},
+            ],
+          ],
+          [
+            101,
+            [
+              {startBlock: 1, endBlock: 300},
+              {startBlock: 50, endBlock: 200},
+            ],
+          ],
+          [201, [{startBlock: 1, endBlock: 300}]],
+          [301, []],
```

Comment:
If this can happen then the indexer will exit because of https://github.com/subquery/subql/blob/5c3142bb41f9196e4fbe433ede8f66021dfbf3cc/packages/node-core/src/indexer/blockDispatcher/base-block-dispatcher.ts#L212-L215

We either need to support this correctly in the fetch service when enqueuing blocks or throw an error if this happens

Reply:
If no datasources are available to index a specific blockheight we throw this error:

https://github.com/subquery/subql/blob/5c3142bb41f9196e4fbe433ede8f66021dfbf3cc/packages/node-core/src/indexer/indexer.manager.ts#L145C1-L159C4

should we change this behaviour to somehow skip blocks that has no datasources?

Reply:
I think that happens too late on having to index to that height to see the error.

I think we can skip blocks, its should be relatively easy to do